File: src/main/java/kome/hatsuneakiko/pastenbt/PasteNbtMod.java
package kome.hatsuneakiko.pastenbt;

import net.minecraft.client.KeyMapping;
import net.minecraftforge.client.event.RegisterClientCommandsEvent;
import net.minecraftforge.client.event.RegisterKeyMappingsEvent;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.event.RegisterCommandsEvent;
import net.minecraftforge.event.TickEvent;
import net.minecraftforge.event.server.ServerAboutToStartEvent;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.event.lifecycle.FMLCommonSetupEvent;
import net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.lwjgl.glfw.GLFW;

@Mod(PasteNbtMod.MODID)
public class PasteNbtMod {
    public static final String MODID = "pastenbt";
    public static final Logger LOGGER = LogManager.getLogger();

    public static final KeyMapping ROTATE_KEY = new KeyMapping("key.pastenbt.rotate", GLFW.GLFW_KEY_R, "key.categories.pastenbt");
    public static final KeyMapping CONFIRM_KEY = new KeyMapping("key.pastenbt.confirm", GLFW.GLFW_KEY_P, "key.categories.pastenbt");
    public static final KeyMapping CANCEL_KEY = new KeyMapping("key.pastenbt.cancel", GLFW.GLFW_KEY_C, "key.categories.pastenbt");

    public PasteNbtMod() {
        FMLJavaModLoadingContext.get().getModEventBus().addListener(this::setup);
        FMLJavaModLoadingContext.get().getModEventBus().addListener(this::registerKeys);
        MinecraftForge.EVENT_BUS.addListener(this::onRegisterCommands);
        MinecraftForge.EVENT_BUS.addListener(this::onRegisterClientCommands);
        MinecraftForge.EVENT_BUS.addListener(this::onServerAboutToStart);
        MinecraftForge.EVENT_BUS.addListener(this::onServerTick);

        // Make the mod server-side optional
        net.minecraftforge.fml.ModLoadingContext.get().registerExtensionPoint(net.minecraftforge.fml.IExtensionPoint.DisplayTest.class, () ->
            new net.minecraftforge.fml.IExtensionPoint.DisplayTest(
                () -> net.minecraftforge.network.NetworkConstants.IGNORESERVERONLY,
                (a, b) -> true
            )
        );
    }

    private void setup(final FMLCommonSetupEvent event) {
        event.enqueueWork(Networking::register);
    }

    private void registerKeys(final RegisterKeyMappingsEvent event) {
        event.register(ROTATE_KEY);
        event.register(CONFIRM_KEY);
        event.register(CANCEL_KEY);
    }

    private void onRegisterCommands(final RegisterCommandsEvent event) {
        NbtPasterCommand.register(event.getDispatcher());
        RtpCommand.register(event.getDispatcher());
        PasteCommand.register(event.getDispatcher());
    }

    private void onRegisterClientCommands(final RegisterClientCommandsEvent event) {
        PasteCommand.register(event.getDispatcher());
    }

    private void onServerAboutToStart(final ServerAboutToStartEvent event) {
        ModConfig.load();
    }

    private void onServerTick(final TickEvent.ServerTickEvent event) {
        if (event.phase == TickEvent.Phase.END) {
            PlacementManager.processTick();
        }
    }
}

File: src/main/java/kome/hatsuneakiko/pastenbt/PasteNbtMod.kt
package kome.hatsuneakiko.pastenbt

import net.minecraft.client.KeyMapping
import net.minecraftforge.client.event.RegisterClientCommandsEvent
import net.minecraftforge.client.event.RegisterKeyMappingsEvent
import net.minecraftforge.common.MinecraftForge
import net.minecraftforge.event.RegisterCommandsEvent
import net.minecraftforge.event.TickEvent
import net.minecraftforge.event.server.ServerAboutToStartEvent
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.event.lifecycle.FMLCommonSetupEvent
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import org.lwjgl.glfw.GLFW
import thedarkcolour.kotlinforforge.forge.FORGE_BUS
import thedarkcolour.kotlinforforge.forge.MOD_BUS

@Mod(PasteNbtMod.MODID)
object PasteNbtMod {
    const val MODID = "pastenbt"
    val LOGGER: Logger = LogManager.getLogger()

    val ROTATE_KEY = KeyMapping("key.pastenbt.rotate", GLFW.GLFW_KEY_R, "key.categories.pastenbt")
    val CONFIRM_KEY = KeyMapping("key.pastenbt.confirm", GLFW.GLFW_KEY_P, "key.categories.pastenbt")
    val CANCEL_KEY = KeyMapping("key.pastenbt.cancel", GLFW.GLFW_KEY_C, "key.categories.pastenbt")

    init {
        MOD_BUS.addListener(::setup)
        MOD_BUS.addListener(::registerKeys)
        FORGE_BUS.addListener(::onRegisterCommands)
        FORGE_BUS.addListener(::onRegisterClientCommands)
        FORGE_BUS.addListener(::onServerAboutToStart)
        FORGE_BUS.addListener(::onServerTick)

        // Làm cho mod trở thành tùy chọn phía Client
        net.minecraftforge.fml.ModLoadingContext.get().registerExtensionPoint(net.minecraftforge.fml.IExtensionPoint.DisplayTest::class.java) {
            net.minecraftforge.fml.IExtensionPoint.DisplayTest(
                { net.minecraftforge.network.NetworkConstants.IGNORESERVERONLY },
                { _, _ -> true }
            )
        }
    }

    private fun setup(event: FMLCommonSetupEvent) {
        event.enqueueWork(Networking::register)
    }

    private fun registerKeys(event: RegisterKeyMappingsEvent) {
        event.register(ROTATE_KEY)
        event.register(CONFIRM_KEY)
        event.register(CANCEL_KEY)
    }

    private fun onRegisterCommands(event: RegisterCommandsEvent) {
        NbtPasterCommand.register(event.dispatcher)
        RtpCommand.register(event.dispatcher)
        PasteCommand.register(event.dispatcher)
    }

    private fun onRegisterClientCommands(event: RegisterClientCommandsEvent) {
        PasteCommand.register(event.dispatcher)
    }

    private fun onServerAboutToStart(event: ServerAboutToStartEvent) {
        ModConfig.load()
    }

    private fun onServerTick(_event: TickEvent.ServerTickEvent) {
        if (_event.phase == TickEvent.Phase.END) {
            PlacementManager.processTick()
        }
    }
}
